modelforge/potential/aimnet2.py

```diff
@@ -168,54 +169,445 @@ def compute_properties(
+            1
+        )  # Shape: (num_atom_pairs, 3, nr_of_features)
+
+        # Sum over the last dimension (nr_of_features) to reduce it
+        vector_prot_step2 = vector_prot_step2.sum(dim=-1)  # Shape: (num_atom_pairs, 3)
+
+        # Initialize tensor to accumulate vector contributions for each atom
+        vector_contributions = torch.zeros(
+            per_atom_feature_tensor.shape[0], 3
+        )  # Shape: (num_of_atoms, 3)
+
+        # Accumulate the vector contributions using index_add_
+        vector_contributions.index_add_(0, idx_j, vector_prot_step2)
+
+        # Step 3: Compute the Euclidean Norm for each atom
+        vector_norms = torch.norm(
+            vector_contributions, p=2, dim=1
+        )  # Shape: (num_of_atoms,)
+
+        return radial_contributions, vector_norms
+
+
+class FirstMessageModule(BaseMessageModule):
+    def __init__(self, number_of_per_atom_features: int):
+        """
+        Initialize the FirstMessageModule.
+
+        Parameters
+        ----------
+        number_of_per_atom_features : int
+            The number of features per atom.
+        """
+        super().__init__(number_of_per_atom_features)
+        self.linear_transform_embeddings = nn.Linear(
+            number_of_per_atom_features, number_of_per_atom_features
+        )
+
+    def forward(
+        self,
+        atomic_embedding: torch.Tensor,
+        partial_charges: torch.Tensor,
+        pairlist: torch.Tensor,
+        f_ij_cutoff: torch.Tensor,
+        r_ij: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Forward pass of the message module.
+
+        Parameters
+        ----------
+        atomic_embedding : torch.Tensor
+            The embedding of each atom.
+        partial_charges : torch.Tensor
+            The partial charges of each atom.
+        pairlist : torch.Tensor
+            The list of atom pairs.
+        f_ij_cutoff : torch.Tensor
+            The cutoff function applied to the radial symmetry functions.
+        r_ij : torch.Tensor
+            The displacement vectors between atom pairs.
+
+        Returns
+        -------
+        torch.Tensor
+            Updated atomic embeddings and partial charges.
+        """
+
+        # Step 1: Calculate radial and vector contributions for atomic embeddings (Equation 4 and 5)
+        radial_contributions_emb, vector_contributions_emb = (
+            self.calculate_contributions(
+                atomic_embedding,
+                pairlist,
+                f_ij_cutoff,
+                r_ij,
+                self.linear_transform_embeddings,
+            )
+        )
+
+        # Step 3: Combine contributions
+        feature_vector = torch.cat(
+            [radial_contributions_emb, vector_contributions_emb.unsqueeze(-1)], dim=1
+        )
+
+        return feature_vector
 
 
-class AimNet2Representation(nn.Module):
+class SubsequentMessageModule(BaseMessageModule):
+    def __init__(self, number_of_per_atom_features: int):
```

Comment:
Like the above constructor, this isn't doing anything that the base constructor isn't already doing---this can be removed. With both of these, it might just make more sense that base constructor is an abstract method.